### ISSUE TITLE
Segmentation fault in ranker fitting

### ISSUE THREAD

**Issue by Keksozavr** [catboost/catboost#1602]
Problem: Segmentation fault
catboost version: python 3.8.7, catboost package version 0.24.4
Operating System: MacOS 10.15.7

Hi everyone!

I'm using catboost as ranking algorithm trained on side-by-side labels.
I wanted to keep good predictive ability and decrease the number of trees (prediction speed is very important to me).
I tried to iterate through some parameters on the grid, but I've caught segmentation fault (in usual training case, I haven't had that issue).

This piece of code should reproduce the error. Sorry for the big constants, it is the easiest way to get segfault.

```
from catboost import CatBoost, Pool
import numpy as np

X_train = np.array([[0.00, 0.00, 0.00, 0.00, 0.18, 0.74, 5.00, 6.00, 10.00, 6.64, 5.86, 5.07, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.00, 0.30, 0.71, 5.00, 9.00, 10.00, 4.79, 4.79, 4.79, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.00, 0.00, -1.00, 5.00, 33.00, 5.00, 6.04, 6.04, 7.99, 0.00, 0.00],
    [1.22, 1.00, 1.22, 0.00, 0.00, -1.00, 1.00, 15.00, 5.00, 3.06, 3.06, 3.42, 1.22, 1.22],
    [1.22, 1.00, 1.22, 0.00, 0.00, -1.00, 1.00, 13.00, 5.00, 5.02, 3.73, 3.74, 1.22, 1.22],
    [1.83, 1.00, 1.04, 0.00, 0.00, -1.00, 2.00, 8.00, 5.00, 5.52, 5.52, 5.55, 1.68, 1.04],
    [0.00, 0.00, 0.00, 0.00, 0.00, -1.00, 4.00, 18.00, 5.00, 3.99, 3.99, 5.76, 0.00, 0.00],
    [1.83, 1.00, 1.04, 0.00, 0.00, -1.00, 2.00, 8.00, 5.00, 5.52, 5.52, 5.55, 1.68, 1.04],
    [0.00, 0.00, 0.00, 0.00, 0.00, -1.00, 1.00, 15.00, 5.00, 3.06, 3.06, 3.42, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.00, 0.00, -1.00, 3.00, 10.00, 5.00, 2.96, 2.84, 2.09, 0.00, 0.00]
])
y_train = np.array([i % 2 for i in range(len(X_train))])
queries_train = ['a', 'a', 'b', 'b', 'b', 'b', 'b', 'b', 'c', 'c']
pairs_train = [(2 * i, 2 * i + 1) for i in range(len(X_train) // 2)]

X_val = np.array([[0.00, 0.00, 0.00, 0.01, 0.34, 0.78, 1.00, 8.00, 15.00, 5.32, 5.26, 5.51, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.00, 0.00, 0.77, 2.00, 8.00, 15.00, 7.48, 3.92, 2.92, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.00, 0.00, 0.30, 4.00, 21.00, 5.00, 5.85, 5.85, 5.98, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.04, 0.02, 0.56, 2.00, 5.00, 5.00, 6.72, 5.72, 3.03, 0.00, 0.00],
    [1.22, 1.00, 1.41, 0.00, 0.23, 0.46, 7.00, 15.00, 15.00, 4.60, 4.17, 4.80, 1.22, 1.39],
    [0.00, 0.00, 2.38, 0.00, 0.20, 0.25, 5.00, 10.00, 15.00, 4.61, 4.61, 5.56, 0.00, 1.88],
    [0.00, 0.00, 0.00, 0.00, 0.05, -0.03, 1.00, 8.00, 10.00, 4.77, 4.67, 5.16, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.02, 0.02, 0.25, 2.00, 12.00, 10.00, 8.04, 6.90, 5.33, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.00, 0.01, 0.37, 2.00, 11.00, 10.00, 4.11, 4.11, 4.54, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.00, 0.30, 0.58, 5.00, 9.00, 10.00, 4.79, 4.79, 4.79, 0.00, 0.00]
])
y_val = np.array([i % 2 for i in range(len(X_val))])
queries_val = ['d', 'd', 'e', 'e', 'f', 'f', 'g', 'g', 'h', 'h']
pairs_val = [(2 * i, 2 * i + 1) for i in range(len(X_val) // 2)]

X_test = np.array([[0.00, 0.00, 0.00, 0.01, 0.34, 0.78, 1.00, 8.00, 15.00, 5.32, 5.26, 5.51, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.00, 0.00, 0.77, 2.00, 8.00, 15.00, 7.48, 3.92, 2.92, 0.00, 0.00],
    [0.00, 0.00, 0.00, 0.19, 0.32, 0.48, 4.00, 11.00, 15.00, 5.81, 5.29, 5.34, 0.00, 0.00],
    [1.22, 1.00, 1.22, 0.24, 0.00, 0.10, 1.00, 6.00, 15.00, 4.23, 3.96, 3.53, 1.22, 1.22],
    [1.22, 1.00, 1.41, 0.00, 0.23, 0.46, 7.00, 15.00, 15.00, 4.60, 4.17, 4.80, 1.22, 1.39],
    [0.00, 0.00, 2.38, 0.00, 0.20, 0.25, 5.00, 10.00, 15.00, 4.61, 4.61, 5.56, 0.00, 1.88],
    [1.83, 1.00, 0.79, 0.00, 0.21, 0.45, 5.00, 33.00, 5.00, 6.04, 6.04, 7.99, 1.68, 0.79],
    [0.00, 0.00, 1.98, 0.00, 0.04, 0.45, 5.00, 11.00, 5.00, 6.26, 6.26, 6.57, 0.00, 1.75],
    [0.00, 0.00, 0.93, 0.00, 0.01, 0.45, 3.00, 3.00, 5.00, 6.58, 6.58, 7.01, 0.00, 0.93],
    [0.00, 0.00, 1.98, 0.00, 0.04, 0.45, 5.00, 11.00, 5.00, 6.26, 6.26, 6.57, 0.00, 1.75]
])
y_test = np.array([i % 2 for i in range(len(X_test))])
queries_test = ['d', 'd', 'f', 'f', 'f', 'f', 'i', 'i', 'i', 'i']
pairs_test = [(2 * i + 1, 2 * i) for i in range(len(X_test) // 2)]

train = Pool(data=X_train, label=y_train, group_id=queries_train, pairs=pairs_train)
val = Pool(data=X_val, label=y_val, group_id=queries_val, pairs=pairs_val)
test = Pool(data=X_test, label=y_test, group_id=queries_test, pairs=pairs_test)

losses = ['QuerySoftMax', 'RMSE', 'QueryRMSE', 'PairLogit', 'PairLogitPairwise']
depths = [4, 6, 8, 10]
iterations = [1, 30, 50, 70, 90, 100]
learning_rates = [0.1, 0.3, 0.5, 0.7, 1.0]

cb_models, cb_accuracy, cb_params = [], [], []
for loss in losses:
    for depth in depths:
        for lr in learning_rates:
            params = {'loss_function': loss, 'iterations': iterations[-1],
                      'learning_rate': lr, 'depth': depth}
            model = CatBoost(params)
            model.fit(train, eval_set=val, verbose=False, plot=False)
            
            for it in iterations[::-1]:
                model.shrink(min([model.tree_count_, it]))
                pred = np.round(model.predict(test))
                cb_models.append(model.copy())
    
                cb_accuracy.append((pred == y_test).sum() / len(y_test))
                params['iterations'] = it
                cb_params.append(params)
```

**Comment by kizill** [comment]
This didn't reproduce neither on Mac, nor on Linux machine. Can you please try running this code under debugger, to help us find the place where problem hides?

**Comment by kizill** [comment]
Closing as irreproducible for now, if you'll help us by collecting crashdumps - we'll continue our investigation :)
